adapter/src/test/java/com/twilio/kudu/sql/PaginationIT.java

```diff
@@ -213,7 +213,8 @@ public void testLimit() throws Exception {
     try (Connection conn = DriverManager.getConnection(url)) {
       String sql = String.format("SELECT * FROM %s LIMIT 3", tableName);
       String expectedPlan = String.format(
-          "KuduToEnumerableRel\n" + "  KuduLimitRel(limit=[3])\n" + "    KuduQuery(table=[[kudu, %s]])\n", tableName);
+          "EnumerableLimit(fetch=[3])\n" + "  KuduToEnumerableRel\n" + "    KuduQuery(table=[[kudu, %s]])\n",
```

Comment:
This means the limit isn't pushed down into the scan.  I think its still worth keeping the KuduLimitRule around for queries that don't have a sort.

Reply:
In the PR description I stated:
>It no longer is derived from EnumerableLimit and
EnumerableLimit is just as efficient as KuduLimit

and I think that deserves some more explanation. [`EnumerableLimit`](https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableLimit.java#L106-L121) uses `ExtendedEnumerable#skip(int)` and `ExtendedEnumerable#take(int)` which is implemented as [`EnumerableDefaults#take(Enumerable<TSource>, int)`](https://github.com/apache/calcite/blob/f30f407654e150679ede6e8d8d4344e6dc26b346/linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java#L3369) and [`EnumerableDefaults#skip(Enumerable<TSource>, int)`](https://github.com/apache/calcite/blob/f30f407654e150679ede6e8d8d4344e6dc26b346/linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java#L3228). These are simple implementations that do what you would expect. 

In [`CalciteEnumerable`](https://github.com/twilio/calcite-kudu/blob/main/adapter/src/main/java/com/twilio/kudu/sql/KuduEnumerable.java#L154-L164), it does the exact same as these two functions -- `EnumerableDefaults#take` and `EnumerableDefaults#skip`. 

> I think its still worth keeping the KuduLimitRule around for queries that don't have a sort.

And therefore, the `KuduLimitRule` doesn't actually add anything and it isn't any more efficient than `EnumerableLimit` -- the number of rows Fetched from Kudu is exactly the same.

Even this [plan](https://github.com/twilio/calcite-kudu/pull/19#discussion_r605052117) should actually not contain a `KuduSortRel` and instead use the `EnumerableLimit`. We should remove `limit` and `offset` from `KuduEnumerable` altogether and simplify that logic within that class.

Reply:
Ah I see. Looking at code more closely we only push down the limit if we are doing a sort
https://github.com/twilio/calcite-kudu/blob/32b5546108abb012e450591e4d3bc9130f342cc1/adapter/src/main/java/com/twilio/kudu/sql/KuduEnumerable.java#L498

However for a query of the form ``SELECT * FROM TABLE LIMIT 10`` we should still be able to push down the limit right ? That would ensure we never send back more than the required number of rows from kudu.

Reply:
I am of the opinion that we shouldn't -- it doesn't make the query more efficient. 

`Enumerator#moveNext` is called for each row, when the [`KuduScanner`](https://github.com/twilio/calcite-kudu/blob/main/adapter/src/main/java/com/twilio/kudu/sql/ScannerCallback.java#L94) has processed all the rows in the first batch it will call this `ScannerCallback#nextBatch`. So this means the number of times `ScannerCallback#nextBatch` is called is exactly the same with `EnumerableLimit` and `KuduLimit` and `KuduSort(fetch, offset and collation)`. The logic we put into `KuduEnumerable` doesn't make the scan more efficient at all.

Reply:
The default batch size is 1MB, so if we don't set a limit the scan will return at least that much data for the call to nextRows even if we only require 1 row for example which will not be as efficient. I think we should take advantage of the push down limit feature since it is supported by kudu and change the logic in KuduEnumerable to push down limits when we don't need the data sorted. 
Impala also utilizes this feature https://gerrit.cloudera.org/#/c/9923/
Here is the original kudu jira https://issues.apache.org/jira/browse/KUDU-16

Reply:
Good call, I think you are suggesting a change to `KuduEnumerable` to:
```java
      if (limit > 0 && !groupBySorted) {
        if (offset > 0) {
           tokenBuilder.limit(offset + limit);
        } else {
           tokenBuilder.limit(limit);
        }
      }
```
> // we can only push down the limit if we are ordering by the pk columns

I don't remember why this required a collation. Can you think of any reason

Reply:
I think this was on oversight on my part, I added the limit push down only if we were required the results sorted and there was no offset.

Reply:
While I am thinking about this, we can't push the `limit` down if the `KuduFilterRel` wasn't able to push all the `Predicate`s into the scan. For instance, `WHERE status <> 'UNKNOWN'`. In this case, there is no Kudu `NOT EQUAL` so we do it in memory. This prevents us from setting the limit on each of the `AsyncKuduScan`s

Reply:
Updated here: https://github.com/twilio/calcite-kudu/pull/20 to fix this up.

Still on the fence with keeping `KuduLimit`, it is harder than I thought to remove it.